### PR TITLE
[opt](Nereids) Support to get notdeterministic functions and  base tables from plan

### PR DESCRIPTION
**Pull request by seawinde**
## Proposed changes

Support to get notdeterministic functions and  base tables from plan.
These are implemented by Plan Visitor

**table collect useage example as following:**
```
TableCollectorContext collectorContext =
    new PlanVisitors.TableCollectorContext(Sets.newHashSet(TableType.OLAP));
physicalPlan.accept(TableCollector.INSTANCE, collectorContext);
```
`collectorContext.getCollectedTables()` has the tables

**nondeterministic functions collect useage example as following:**
```
List<Expression> functionCollectResult = new ArrayList<>();
plan.accept(NondeterministicCollector.INSTANCE, collectResult);
```
`functionCollectResult` has the nondeterministic functions

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

**Review by zddr — APPROVED** [review]
LGTM

**Comment by seawinde** [comment]
run buildall

**Comment by seawinde** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/TreeNode.java [hunk 1]
```diff
@@ -184,6 +184,27 @@ default boolean anyMatch(Predicate<TreeNode<NODE_TYPE>> predicate) {
...
+     * iterate top down and test predicate, if matched then collect it.
+     *
+     * @param predicate predicate
+     * @param matchedNodes the nodes that satisfies the predicate
+     * @param traversalAll whether continue to recurse or not when predicate is true
+     */
+    default void matchAndCollect(Predicate<TreeNode<NODE_TYPE>> predicate,
+            List<TreeNode<NODE_TYPE>> matchedNodes,
+            boolean traversalAll) {
+        if (predicate.test(this)) {
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/TreeNode.java]: It's the same, will use treeNode collect method instead

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+     * Collect the nondeterministic expr in plan, these expressions will be put into context
+     */
+    public static class NondeterministicFunctionCollector
+            extends DefaultPlanVisitor<Void, List<TreeNode<Expression>>> {
+        @Override
+        public Void visit(Plan plan, List<TreeNode<Expression>> collectedExpressions) {
+            List<? extends Expression> expressions = plan.getExpressions();
+            if (expressions == null) {
+                return super.visit(plan, collectedExpressions);
+            }
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java]: modify it to `isEmpty()` test

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+            expressions.forEach(expression -> expression.matchAndCollect(
+                    expressionNode -> expressionNode instanceof Nondeterministic,
```
[the remark this answers is not in the document]
Comment (seawinde) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/visitor/PlanVisitors.java]: OK, have fixed it

**Review by morrySnow — APPROVED** [review]

**Pull request merged**